cpp/src/svm/linear.cu

```diff
@@ -0,0 +1,462 @@
+                  params.max_iter,
+                  T(params.grad_tol),
+                  T(params.change_tol),
+                  params.linesearch_max_iter,
+                  params.lbfgs_memory,
+                  params.verbose,
+                  w1,
+                  &target,
+                  &num_iters,
+                  qn_loss,
+                  stream,
+                  (T*)sampleWeight,
+                  T(params.svr_sensitivity));
+
+    if (coefCols > 1)
+      rowMajorSetCol<T><<<dim3(raft::ceildiv(coefRows, 256), 1, 1), dim3(256, 1, 1), 0, stream>>>(
+        w.data(), w1, class_i, coefRows, coefCols);
+
+    CUDA_CHECK(cudaStreamSynchronize(stream));
+    CUML_LOG_DEBUG(
+      "LinearSVM finished fitting in %d iterations out of maximum %d.", num_iters, params.max_iter);
+  }
+
+  ML::POP_RANGE();
+
+  /** TODO: probabolisting calibration is disabled for now, multiclass case is not ready. */
+  // if (!params.probability) return;
+
+  // ML::PUSH_RANGE("Trace::LinearSVMModel::fit-probabilities");
+  // probScale.resize(coefCols * (coefCols + 1), stream);
+
+  // rmm::device_uvector<T> xwBuf(nRows * coefCols, stream);
+  // T* xw = xwBuf.data();
+  // predict_linear(handle, X, w.data(), nRows, nCols, coefCols, params.fit_intercept, xw, stream);
+
+  // // here we should encode the labels into corresponding `classes` indices.
+  // // for now, we assume labels are the values from 0 to classes.size() - 1.
+  // // raft::linalg::unaryOp(y1, y, nRows, IndicatorFun<T>{T(params.H1_value)}, stream);
+
+  // GLM::qnFit<T>(handle,
+  //               xw,
+  //               false,
+  //               (T*)y,
+  //               nRows,
+  //               coefCols,
+  //               nClasses,
+  //               true,
+  //               0,
+  //               0,
+  //               params.max_iter,
+  //               T(params.grad_tol),
+  //               T(params.change_tol),
+  //               params.linesearch_max_iter,
+  //               params.lbfgs_memory,
+  //               params.verbose,
+  //               probScale.data(),
+  //               &target,
+  //               &num_iters,
+  //               nClasses == 2 ? 0 /* logistic loss*/ : 2 /** softmax */,
+  //               stream,
+  //               (T*)sampleWeight);
+
+  // CUML_LOG_DEBUG("LinearSVM finished fitting probabilities in %d iterations out of maximum %d.",
+  //                num_iters,
+  //                params.max_iter);
+
+  // ML::POP_RANGE();
+}
+
+template <typename T>
+void LinearSVMModel<T>::predict(const T* X, const int nRows, const int nCols, T* out) const
+{
+  cudaStream_t stream = handle.get_stream();
+  const int nClasses  = classes.size();
+  const int coefCols  = nClasses <= 2 ? 1 : nClasses;
+  if (isRegression(params.loss))
+    return predict_linear(
+      handle, X, w.data(), nRows, nCols, coefCols, params.fit_intercept, out, stream);
+
+  rmm::device_uvector<T> temp(nRows * coefCols, stream);
```

Comment:
Not really, because the next step (as it is now) cannot be performed in-place. `predict_proba` returns `nClasses` probabilities for each input (even in the binary case), whereas `temp` in the binary case contains a single score per input.